file.go

```diff
@@ -82,7 +82,12 @@ func (f *file) Write(p []byte) (int, error) {
 		return 0, ErrReadOnlyFile
 	}
 
-	return f.w.Write(p)
+	n, err := f.w.Write(p)
+	if err == siva.ErrMissingHeader {
+		return 0, os.ErrClosed
```

Comment:
Maybe you're stuck with this because of compatibility, but this seems like it could be really hard to debug.

```go
if _, err := s.Write(bits); err != nil {
  return fmt.Errorf("writing to siva: %v", err)
}
```

further up the stack, `main` calls `log.Fatalf("Creating archive: %v", err)` and now I get a log that looks like

    2019-07-31 13:26:52.1029330 Creating archive: writing to siva: file already closed

So now instead of going after the actual problem, which is that I forgot to write a header (I think), I'm going to go looking for concurrency bugs in my file descriptor handling.

Reply:
Maybe we can add a `TODO` to change the behavior in the next version of `go-billy-siva` but the behavior before `go-siva` change was returning `os.ErrClosed` after calling sync. Here's an excerpt of the old test:

```go
	f, err := s.FS.Create("testOne.txt")
	c.Assert(err, IsNil)

	n, err := f.Write([]byte("qux"))
	c.Assert(err, IsNil)
	c.Assert(n, Equals, 3)

	err = s.FS.Sync()
	c.Assert(err, IsNil)

	n, err = f.Write([]byte("bar"))
	c.Assert(err.(*os.PathError).Err, Equals, os.ErrClosed)
	c.Assert(n, Equals, 0)
```

I'm not sure if there's any big implication changing this behavior. AFAIK there's no code from us checking if the file was already closed but I stayed on the safe side and maintained the error.

As a side note I changed the test to compare directly with `os.ErrClosed` as it's no longer an `os.PathError` in newer go versions.

Reply:
maybe as a future improvement we should wrap errors and check the chain of them

Reply:
I doubt this change will break any existing code, my comment was more about debugging any errors that come from it. It theoretically possible for this to break something, but it's not common to check for "already closed" type errors separately from others, so I wouldn't worry about that too much:

```go
_, err := s.Write(someData)
if err == io.ErrClosed {
  return errors.New("holy crap something is really wrong")
} else if err := s.Close(); err != nil {
  return fmt.Errorf("closing: %v", err)
}
```

Anyway, not to derail your current change, just something to consider for future work.